src/client/storage/index.js

```diff
@@ -68,32 +77,56 @@ const getChunk = async (chunkId, encoding = 'utf8', useCache = true) => {
         log.debug({chunkId}, 'Graphql request success');
 
         for (const edge of queryResult.transactions.edges) {
```

Comment:
What I think, the flow is a bit redundant.
Here's how we download chunks from arweave: by chunk id we fetch graphql with tx ids array, than loop through this array trying to find data, which hash matches the id. In theory, there should not be any data with wrong id, but I think it can happen, maybe if data is broken. Also, it can happen for some reason that the array is empty.
So i think you should do the fallback s3 fetching not inside this loop - it actually doesn't make much sense, since s3 doesn't rely on tx id at all.
Instead, this fallback should be lower, in a catch block - so if arweave fails for any reason (graphql failed, or graphql res empty, or all of the nodes failed) - we fall back to s3

Reply:
I suggest we tackle this one separately. To have a hotfix out as soon as possible. There are many possibilities for improvement throughout the whole process.